batch/WFLY-15525_Limit_number_of_records_loaded_by_jdbc_store.adoc

```diff
@@ -0,0 +1,145 @@
+failed deployments were also reported.
+
+Several fixes have already been completed, which optimized processing of these large job repositories:
+
+* https://issues.redhat.com/browse/WFLY-14275
+* https://issues.redhat.com/browse/WFLY-14946
+
+While these fixes improved the deployment times considerably, an upstream user reported that application deployments still take
+significantly longer in their environment, than they used to take with an older version of Wildfly.
+
+Debugging performed by us and by the upstream user indicates that most time is spent on reading records from the JDBC ResultSet, not
+on executing the actual database query. Database latency seems to be an important factor.
+
+Another issue noticed later on is that the Wildfly Web Console is not able to process and display such large number job executions,
+as the operation that loads the data fails with OutOfMemory exception. (Neither is the Web Console designed to display that amount of
+records.)
+
+Hence, it is proposed to implement a way to limit the number of execution records that are loaded by the job repository
+implementations.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-15525
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/WFLY-14275
+* https://issues.redhat.com/browse/WFLY-14946
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+* [x] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+* JBeret Core,
+* Batch subsystem.
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+Make it possible to limit the number of job executions returned by job repository implementations. If this limitation behavior
+is active, the subset of records returned by the job repository must contain the youngest records, sorted from the youngest to
+the oldest. Older records which do not fit into the specified limit will be omitted.
+
+If the limit is not specified, current behavior will apply and all available records will be loaded.
+
+=== Nice-to-Have Requirements
+
+* If the limiting behavior comes to an effect, print a debug or a warning log message.
+* Add debug messages that would make it easier to identify where the JDBC job repository spends time when loading execution
+records.
+
+//=== Non-Requirements
+
+== Implementation Plan
+
+JBeret and Wildfly code bases need to be modified.
+
+=== JBeret Core
+
+Currently, the JobRepository interface has method `List<Long> getJobExecutionsByJob(String jobName)`, which returns a list
+of all job execution IDs for given jobName present in the repository.
+
+The proposal is to add a new method `List<Long> getJobExecutionsByJob(String jobName, Long maxRecords)`, that would return
+a list with maximum size of `maxRecords`.
```

Comment:
Did you mean `int maxRecords`?

What's the range of this attribute? min = 1, max = some reasonably big int?

Reply:
I actually wanted `Integer maxRecords`, where null value would mean no limit. It would make implementation little bit simpler, as the original method could just delegate to `getJobExecutionsByJob(jobName, null)`.

I think that min = 1 is OK, but not sure if we should set max. If the user wants to be able to list executions in Web Console, he should stick to couple of hundreds max. If the user wants to be able to load executions via CLI (and e.g. process with `grep`), much higher limits are OK. So I don't know where to draw the line, and for that reason I would not define max.